server/src/test/java/org/opensearch/indices/replication/RemoteStoreReplicationSourceTests.java

```diff
@@ -177,6 +181,77 @@ public void testGetCheckpointMetadataEmpty() throws ExecutionException, Interrup
+                    if (segment.startsWith("segments_") == false) put(segment, metadata.getUploadedFilename());
+                });
+            }
+        };
+        replicaShard.getRemoteDirectory().markPendingMergedSegmentsDownload(localToRemoteFilenameMap);
+        RemoteStoreMergedSegmentCheckpoint mergedSegmentCheckpoint = new RemoteStoreMergedSegmentCheckpoint(
+            new MergedSegmentCheckpoint(
+                replicaShard.shardId(),
+                primaryTerm,
+                checkpoint.getLength(),
+                checkpoint.getCodec(),
+                checkpoint.getMetadataMap(),
+                "_0"
+            ),
+            localToRemoteFilenameMap
+        );
+        List<StoreFileMetadata> filesToFetch = primaryShard.getSegmentMetadataMap().values().stream().collect(Collectors.toList());
+        final PlainActionFuture<GetSegmentFilesResponse> res = PlainActionFuture.newFuture();
+        replicationSource = new RemoteStoreReplicationSource(primaryShard);
+        replicationSource.getMergedSegmentFiles(
+            REPLICATION_ID,
+            mergedSegmentCheckpoint,
+            filesToFetch,
+            replicaShard,
+            (fileName, bytesRecovered) -> {},
+            res
+        );
+        GetSegmentFilesResponse response = res.get();
+        assertEquals(response.files.size(), filesToFetch.size());
+        assertTrue(response.files.containsAll(filesToFetch));
+        closeShards(replicaShard);
+    }
+
+    public void testGetMergedSegmentFilesFailure() throws IOException {
+        // Testing failure scenario where segments are not a part of RemoteSegmentStoreDirectory.pendingMergedSegmentsDownloads
+        final ReplicationCheckpoint checkpoint = primaryShard.getLatestReplicationCheckpoint();
+        Map<String, String> localToRemoteFilenameMap = Map.of("invalid.si", "invalid.si__uuid", "invalid.cfs", "invalid.cfs__uuid");
+        RemoteStoreMergedSegmentCheckpoint mergedSegmentCheckpoint = new RemoteStoreMergedSegmentCheckpoint(
+            new MergedSegmentCheckpoint(
+                replicaShard.shardId(),
+                primaryTerm,
+                checkpoint.getLength(),
+                checkpoint.getCodec(),
+                checkpoint.getMetadataMap(),
+                "_0"
+            ),
+            localToRemoteFilenameMap
+        );
+        List<StoreFileMetadata> filesToFetch = primaryShard.getSegmentMetadataMap().values().stream().collect(Collectors.toList());
+        final PlainActionFuture<GetSegmentFilesResponse> res = PlainActionFuture.newFuture();
+        replicationSource = new RemoteStoreReplicationSource(primaryShard);
+        replicationSource.getMergedSegmentFiles(
+            REPLICATION_ID,
+            mergedSegmentCheckpoint,
+            filesToFetch,
+            replicaShard,
+            (fileName, bytesRecovered) -> {},
+            res
+        );
+        assertThrows(ExecutionException.class, () -> res.get());
```

Comment:
can we also test the expected exception message? Also, should we cover any more failure scenarios?